caas/kubernetes/provider/customresourcedefinitions.go

```diff
@@ -82,3 +107,224 @@ func (k *kubernetesClient) deleteCustomResourceDefinitions(appName string) error
+	logger.Debugf("updating custom resource %q", cr.GetName())
+	// TODO(caas): do label check to ensure the resource to be updated was created by Juju once caas upgrade steps of 2.7 in place.
+	out, err = api.Update(cr)
+	return out, cleanUps, errors.Trace(err)
+}
+
+func deleteCustomResourceDefinition(api dynamic.ResourceInterface, name string, uid types.UID) error {
+	err := api.Delete(name, newPreconditionDeleteOptions(uid))
+	if k8serrors.IsNotFound(err) {
+		return nil
+	}
+	return errors.Trace(err)
+}
+
+func (k *kubernetesClient) getCRDsForCRs(crs map[string][]unstructured.Unstructured) (out map[string]*apiextensionsv1beta1.CustomResourceDefinition, err error) {
+	var wg sync.WaitGroup
+	wg.Add(len(crs))
+	out = make(map[string]*apiextensionsv1beta1.CustomResourceDefinition)
+	crdChan := make(chan *apiextensionsv1beta1.CustomResourceDefinition, 1)
+	errChan := make(chan error, 1)
+
+	getCRD := func(
+		name string,
+		broker *kubernetesClient,
+		resultChan chan<- *apiextensionsv1beta1.CustomResourceDefinition,
+		errChan chan<- error,
+		clk jujuclock.Clock,
+	) {
+		checker := func() error {
+			crd, err := broker.getCustomResourceDefinition(name, false)
+			if err != nil {
+				return errors.Annotatef(err, "getting custom resource definition %q", name)
+			}
+			version := crd.Spec.Version
+			if version == "" {
+				if len(crd.Spec.Versions) == 0 {
+					return errors.NotValidf("custom resource definition %q no version", crd.GetName())
+				}
+				version = crd.Spec.Versions[0].Name
+			}
+			crClient, err := broker.getCustomResourceDefinitionClient(crd, version)
+			if err != nil {
+				return errors.Annotatef(err, "getting custom resource definition client %q", name)
+			}
+			if _, err := crClient.List(v1.ListOptions{IncludeUninitialized: false}); err != nil {
+				if k8serrors.IsNotFound(err) {
+					// CRD already exists, but the resource type does not exist yet.
+					return errors.NewNotFound(err, fmt.Sprintf("custom resource definition %q resource type", crd.GetName()))
+				}
+				return errors.Trace(err)
+			}
+			resultChan <- crd
+			return nil
+		}
+
+		err := retry.Call(retry.CallArgs{
+			Attempts: 8,
+			Delay:    1 * time.Second,
+			Clock:    clk,
+			Func:     checker,
+			IsFatalError: func(err error) bool {
+				return err != nil && !errors.IsNotFound(err)
+			},
+			NotifyFunc: func(err error, attempt int) {
+				logger.Debugf("fetching custom resource definition %q, err %#v, attempt %v", name, err, attempt)
+			},
+		})
+		if err != nil {
+			errChan <- err
+		} else {
+			wg.Done()
+		}
+	}
+
+	for name := range crs {
+		go getCRD(name, k, crdChan, errChan, k.clock)
+	}
+
+	done := make(chan struct{})
+	go func() {
+		wg.Wait()
+		close(done)
+	}()
+
+	for {
+		select {
+		case crd := <-crdChan:
+			out[crd.GetName()] = crd
+		case <-done:
+			return out, nil
+		case err := <-errChan:
+			return out, errors.Annotatef(err, "getting custom resources")
+		case <-k.clock.After(10 * time.Second):
```

Comment:
is this enough time for slow clouds?

can we construct the retry.Call() above in terms of the timeout here, eg the time specified here minus 1s is used for retry arg

Reply:
changed to `15s`, and added `Stop` chan to the `retry`